addons/mail/static/src/new/core/chat_window_model.js

```diff
@@ -0,0 +1,33 @@
+/** @odoo-module */
+
+export class ChatWindow {
+    /**
+     * @param {import("@mail/new/core/messaging").Messaging['state']} state
+     * @param {Object} data
+     * @returns {ChatWindow}
+     */
+    static insert(state, data) {
+        const { threadId } = data;
+        const chatWindow = state.chatWindows.find((c) => c.threadId === threadId);
+        if (!chatWindow) {
+            const chatWindow = new ChatWindow(data);
+            chatWindow._state = state;
+            state.chatWindows.push(chatWindow);
+        } else {
+            chatWindow.folded = false;
+            chatWindow.autofocus++;
+        }
+    }
+
+    constructor(data) {
+        const { threadId } = data;
+        Object.assign(this, { threadId, autofocus: 1, folded: false });
+    }
```

Comment:
It would be better to not have constructor and have insert actually handle updating data.